src/allocation_engine/interval_tree.rs

```diff
@@ -286,4 +451,158 @@ mod tests {
         ns = NodeState::Free;
         assert!(ns.is_free());
     }
+
+    #[test]
+    fn test_search() {
+        let left_child = InnerNode::new(Range::new(0x90, 0x99).unwrap(), NodeState::Free);
+        let right_child = InnerNode::new(Range::new(0x200, 0x2ff).unwrap(), NodeState::Free);
+        let mut root_node = InnerNode::new(Range::new(0x100, 0x110).unwrap(), NodeState::Allocated);
+        root_node.left = Some(Box::new(left_child.clone()));
+        root_node.right = Some(Box::new(right_child));
+
+        assert_eq!(
+            root_node.search(&Range::new(0x90, 0x99).unwrap()),
+            Some(&left_child)
+        );
+        assert_eq!(root_node.search(&Range::new(0x200, 0x250).unwrap()), None);
+        assert_eq!(root_node.search(&Range::new(0x111, 0x1fe).unwrap()), None);
+    }
+
+    #[test]
+    fn test_search_superset() {
+        let left_child = InnerNode::new(Range::new(0x90, 0x99).unwrap(), NodeState::Free);
+        let right_child = InnerNode::new(Range::new(0x200, 0x2ff).unwrap(), NodeState::Free);
+        let mut root_node = InnerNode::new(Range::new(0x100, 0x110).unwrap(), NodeState::Allocated);
+        root_node.left = Some(Box::new(left_child));
+        root_node.right = Some(Box::new(right_child.clone()));
+
+        assert_eq!(
```

Comment:
Can you write a bit about the test cases here? There are a lot of asserts and it's a bit hard to follow.